### PR TITLE
tracing: add support for tracing db interactions

### PR DESCRIPTION
**Pull request by aaslamin**
This PR adds support for creating spans around DB interactions by leveraging our effort in https://github.com/ory/x/pull/3 (_which utilizes the library created by @luna-duclos to decorate the interfaces exposed by the `database/sql` package with the opentracing API_).

#### Things of interest:

- An instrumented driver will only be used IFF tracing has been configured
  - Perhaps we should add a config specifically for creating spans around DB interactions incase someone does not care about such spans? 
- ~Configuration added to omit SQL args from being added to spans~
- New test coverage to exercise new tracing options

#### Example trace from making a bad call to `POST /clients`

image
---

_Review: @aeneasr_

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -216,6 +216,9 @@ func initConfig() {
 	viper.BindEnv("TRACING_SERVICE_NAME")
 	viper.SetDefault("TRACING_SERVICE_NAME", "Ory Hydra")
 
+	viper.BindEnv("TRACING_OMIT_SQL_ARGUMENTS_FROM_SPANS")
```
Comment (aeneasr) [inline cmd/root.go]: Please document this around here: https://github.com/ory/hydra/blob/master/cmd/serve.go#L182

And while you're add it, please document the other tracing env vars there too :)

File: config/backend_connector.go [hunk 2]
```diff
@@ -41,7 +41,7 @@ var (
 )
 
 type BackendConnector interface {
-	Init(url string, l logrus.FieldLogger) error
+	Init(url string, l logrus.FieldLogger, opts ...ConnectorOptions) error
```
Comment (aeneasr) [inline config/backend_connector.go]: ping @someone1 fyi

File: config/connector_options.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package config
+
+type ConnectorOptions func(*options)
+
+type options struct {
```
Comment (aeneasr) [inline config/connector_options.go]: would it make sense to reuse those from `sqlcon`?

Comment (aeneasr) [reply config/connector_options.go]: That makes sense to me

**Comment by aaslamin** [comment]
Thanks for the feedback @aeneasr @someone1, looking into them now 🙌

File: config/connector_options.go (hunk 3 above)
Comment (aaslamin) [reply config/connector_options.go]: Re: `OmitSQLArgsFromTracingSpans` I was following some of the concerns that Google identified in [Dapper](https://ai.google/research/pubs/pub36356) for not releasing potentially sensitive details in traces. Perhaps in some organizations, the engineers that have access to look at traces should not be able to view such data for compliance/privacy reasons.

I can remove this feature until we get a feature request, maybe?

Comment (someone1) [reply config/connector_options.go]: My apologies for not being clear on that - I meant that traces should always omit the data. For general hydra use-cases, I don't think there's a time this is needed outside of debugging/testing.

Comment (aaslamin) [reply config/connector_options.go]: Sweet, that sounds good to me. I will leave this option on by default and remove the config and if people ask for it be included, they can create an issue so we can re-introduce it.

File: config/backend_connector.go (hunk 2 above)
Comment (aaslamin) [reply config/backend_connector.go]: Should be exported now. I was completely unaware of the plugin use case, exported options should be generic enough for any type of connector, not just SQL (e.g. redis, mongo, etc).

File: config/config.go [hunk 4]
```diff
@@ -317,22 +327,20 @@ func (c *Config) Context() *Context {
 	}
 
 	if backend, ok := backends[scheme]; ok {
-		if err := backend.Init(c.DatabaseURL, c.GetLogger()); err != nil {
+		options := []ConnectorOptions{}
+
+		if tracingEnabled {
+			options = append(options, WithTracing(), withOmitSQLArgsFromSpans())
```
Comment (aaslamin) [inline config/config.go]: I could have done something like `os.Getenv("OMIT_SQL_ARGS_FROM_TRACES")` to give _advanced users_ the option, but I felt dirty doing it? ¯\_(ツ)_/¯

File: config/connector_options.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package config
+
+type ConnectorOptions func(*Options)
+
+type Options struct {
```
Comment (aaslamin) [inline config/connector_options.go]: 👀

File: config/connector_options.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	UseTracing bool
+
+	// unexported as these are specifically used by Hydra for writing tests
```
Comment (aaslamin) [inline config/connector_options.go]: 👀

File: config/config.go (hunk 4 above)
Comment (aeneasr) [reply config/config.go]: Yeah, let's just keep it as is. If people complain we can add it or they have to re-build the binary which hopefully makes them incredibly aware of removing that once they're done.

File: config/connector_options.go (hunk 6 above)
Comment (aeneasr) [reply config/connector_options.go]: This makes a ton of sense to me, all of these options are only used internally for testing and other stuff. Good job!

Comment (aaslamin) [reply config/connector_options.go]: Yes, that was completely my bad. I was not aware of the plugin use case. It makes sense to me 👍

File: config/config.go [hunk 7]
```diff
@@ -294,6 +294,16 @@ func (c *Config) Context() *Context {
...
 	var hasher fosite.Hasher
+	var tracingEnabled bool
+
+	hasher = &fosite.BCrypt{
+		WorkFactor: c.BCryptWorkFactor,
+	}
+
+	if tracer, err := c.GetTracer(); err == nil && tracer.IsLoaded() {
+		tracingEnabled = true
+		hasher = &tracing.TracedBCrypt{c.BCryptWorkFactor}
```
Comment (aeneasr) [inline config/config.go]: Please use keyed fields here

Comment (aeneasr) [reply config/config.go]: I also noticed that this is kinda duplicated with [here](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L120-L122). I also noticed that [`hasher`](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L56) might be null if [tracing is disabled](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L120-L122). So instead this should use the bcrypt from the config (here)

Comment (aeneasr) [reply config/config.go]: The nil thing is actually default behavior and did not cause any issues. `BCryptWorkFactor` was also set properly.

Comment (aeneasr) [reply config/config.go]: I think for clarity it should we should use the hasher from the config [here](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L120-L122).

Comment (aaslamin) [reply config/config.go]: > I also noticed that this is kinda duplicated with [here](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L120-L122). I also noticed that [`hasher`](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L56) might be null if [tracing is disabled](https://github.com/ory/hydra/blob/master/cmd/server/handler_oauth2_factory.go#L120-L122). So instead this should use the bcrypt from the config (here)

If Hasher is `nil` it gets set by fosite when Compose is called https://github.com/ory/fosite/blob/master/compose/compose.go#L56

Comment (aaslamin) [reply config/config.go]: 👍 on DRYing up the check to see if tracing has been loaded tho. I will clean that up.

**Comment by someone1** [comment]
LGTM

Great job!

File: tracing/tracer.go [hunk 8]
```diff
@@ -76,3 +76,43 @@ func (t *Tracer) Close() {
...
+- TRACING_PROVIDER_JAEGER_SAMPLING_VALUE: The value passed to the sampler type that has been configured.
+
+	Supported values: This is dependant on the sampling strategy used:
+		- const: 0 or 1 (all or nothing)
+		- rateLimiting: a constant rate (e.g. setting this to 3 will sample requests with the rate of 3 traces per second)
+		- probabilistic: a value between 0..1
+
+	Example: TRACING_PROVIDER_JAEGER_SAMPLING_VALUE=1
+
+- TRACING_PROVIDER_JAEGER_LOCAL_AGENT_HOST_PORT: The address of the jaeger-agent where spans should be sent to
```
Comment (aeneasr) [inline tracing/tracer.go]: Let's call this `TRACING_PROVIDER_JAEGER_LOCAL_AGENT_ADDRESS`

File: tracing/tracer.go [hunk 9]
```diff
@@ -76,3 +76,43 @@ func (t *Tracer) Close() {
...
+
+	Example: TRACING_PROVIDER_JAEGER_LOCAL_AGENT_HOST_PORT=127.0.0.1:6831
+
+- TRACING_SERVICE_NAME: Specifies the service name to use on the tracer.
+
+	Default: Ory Hydra
```
Comment (aeneasr) [inline tracing/tracer.go]: `Ory` should be uppercase: `ORY`

**Comment by aeneasr** [comment]
Thank you for your hard work!

**Pull request merged**